intercept/js/server_connector.js

```diff
@@ -0,0 +1,182 @@
+  // promise may be rejected if there are connectivity issues, a device with
+  // that id doesn't exist or this client doesn't have rights to access that
+  // device.
+  async requestDevice(deviceId) {
+    throw 'Not implemented!';
+  }
+
+  // Sends a message to the device selected with requestDevice. It's an error to
+  // call this function before the promise from requestDevice() has resolved.
+  // Returns an empty promise that is rejected when the message can not be
+  // delivered, either because the device has not been requested yet or because
+  // of connectivity issues.
+  async sendToDevice(msg) {
+    throw 'Not implemented!';
+  }
+}
+
+// End of Server Connector Interface.
+
+// The following code is internal and shouldn't be accessed outside this file.
+
+function httpUrl(path) {
+  return location.protocol + '//' + location.host + '/' + path;
+}
+
+async function ajaxPostJson(url, data) {
+  const response = await fetch(url, {
+    method: 'POST',
+    cache: 'no-cache',
+    headers: {'Content-Type': 'application/json'},
+    redirect: 'follow',
+    body: JSON.stringify(data),
+  });
+  return response.json();
+}
+
+// Implementation of the Connector interface using HTTP long polling
+class PollingConnector extends Connector {
+  #configUrl = httpUrl('infra_config');
+  #connectUrl = httpUrl('connections');
+  #forwardUrl;
+  #messagesUrl;
+  #config = undefined;
+  #messagesReceived = 0;
+  #pollerSchedule;
+  #onDeviceMsgCb = msg =>
+      console.error('Received device message without registered listener');
+
+  onDeviceMsg(cb) {
+    this.#onDeviceMsgCb = cb;
+  }
+
+  constructor() {
+    super();
+  }
+
+  async requestDevice(deviceId) {
+    let config = await this.#getConfig();
+    let response = await ajaxPostJson(this.#connectUrl, {device_id: deviceId});
+    let connId = response.connection_id;
+    this.#forwardUrl = httpUrl(`connections/${connId}/:forward`);
+    this.#messagesUrl = httpUrl(`connections/${connId}/messages`);
+
+    this.#startPolling();
+
+    return {
+      deviceInfo: response.device_info,
+      infraConfig: config,
+    };
+  }
+
+  async sendToDevice(msg) {
+    return await ajaxPostJson(this.#forwardUrl, {
+      payload: msg,
+    });
+  }
+
+  async #getConfig() {
+    if (this.#config === undefined) {
+      this.#config = await (await fetch(this.#configUrl, {
+                       method: 'GET',
+                       redirect: 'follow',
+                     })).json();
+    }
+    return this.#config;
+  }
+
+  async #pollMessages() {
+    let r = await fetch(
+        this.#messagesUrl + `?start=${this.#messagesReceived}`, {
+          method: 'GET',
+          redirect: 'follow',
+        })
+    let arr = await r.json();
+    this.#messagesReceived += arr.length;
+    return arr;
+  }
+
+  #startPolling() {
+    if (this.#pollerSchedule !== undefined) {
+      return;
+    }
+
+    let currentPollDelay = 1000;
```

Comment:
should we declare a constant for the initial poll delay?

Reply:
we could, but since it won't be used anywhere else and 1000 is quite commonly used as 1 second in milliseconds, especially in javascript, I thought it wouldn't make the code much clearer than just using the number directly.